core/src/main/scala/kafka/log/Log.scala

```diff
@@ -893,6 +926,7 @@ class Log(@volatile var dir: File,
         activeSegment.truncateTo(targetOffset)
         updateLogEndOffset(targetOffset)
         this.recoveryPoint = math.min(targetOffset, this.recoveryPoint)
+        this.logStartOffset = math.min(targetOffset, this.logStartOffset)
```

Comment:
Discussed offline. Given that Log.truncateTo(...) has explicitly consider the case that the targetOffset is smaller than base offset of the first log segment, it would be safer and simpler for us to assume that Log.truncateTo(...) can truncate to an offset smaller than logStartOffset.